src/coreclr/nativeaot/Runtime/interoplibinterface_objc.cpp

```diff
@@ -0,0 +1,179 @@
+
+#include "common.h"
+#include "CommonTypes.h"
+#include "CommonMacros.h"
+#include "daccess.h"
+#include "PalRedhawkCommon.h"
+#include "PalRedhawk.h"
+#include "rhassert.h"
+#include "slist.h"
+#include "holder.h"
+#include "gcrhinterface.h"
+#include "shash.h"
+#include "RWLock.h"
+#include "rhbinder.h"
+#include "Crst.h"
+#include "RuntimeInstance.h"
+#include "TypeManager.h"
+#include "MethodTable.h"
+#include "ObjectLayout.h"
+#include "event.h"
+#include "varint.h"
+#include "regdisplay.h"
+#include "StackFrameIterator.h"
+#include "thread.h"
+#include "threadstore.h"
+#include "threadstore.inl"
+
+#include "interoplibinterface.h"
+
+#include "MethodTable.inl"
+
+#ifdef FEATURE_OBJCMARSHAL
+
+namespace
+{
+    // TODO: Support registering multiple begin/end callbacks.
+    // NativeAOT support the concept of multiple managed "worlds",
+    // each with their own object heirarchy. Each of these worlds
+    // could potentially have its own set of callbacks.
+    // However this Objective-C Marshal API was not designed with such a
+    // possibility in mind.
+    ObjCMarshalNative::BeginEndCallback s_pBeginEndCallback = nullptr;
+
+    struct DisableTriggerGcWhileCallingManagedCode
+    {
+        Thread * m_pThread;
+        bool m_fGcStressWasSuppressed;
+
+        DisableTriggerGcWhileCallingManagedCode()
+        {
+            // It is illegal for any of the callouts to trigger a GC.
+            m_pThread = ThreadStore::GetCurrentThread();
+            m_pThread->SetDoNotTriggerGc();
+
+            // Due to the above we have better suppress GC stress.
+            m_fGcStressWasSuppressed = m_pThread->IsSuppressGcStressSet();
+            if (!m_fGcStressWasSuppressed)
+                m_pThread->SetSuppressGcStress();
+        }
+
+        ~DisableTriggerGcWhileCallingManagedCode()
+        {
+            // Revert GC stress mode if we changed it.
+            if (!m_fGcStressWasSuppressed)
+                m_pThread->ClearSuppressGcStress();
+
+            m_pThread->ClearDoNotTriggerGc();
+        }
+    };
+
+    bool TryGetTaggedMemory(_In_ Object * obj, _Out_ void ** tagged)
+    {
+        void* fn = obj->get_EEType()
+                       ->GetTypeManagerPtr()
+                       ->AsTypeManager()
+                       ->GetClasslibFunction(ClasslibFunctionId::ObjectiveCMarshalTryGetTaggedMemory);
+
+        ASSERT(fn != nullptr);
+
+        auto pTryGetTaggedMemoryCallback = reinterpret_cast<ObjCMarshalNative::TryGetTaggedMemoryCallback>(fn);
+
+        DisableTriggerGcWhileCallingManagedCode disabler{};
+
+        bool result = pTryGetTaggedMemoryCallback(obj, tagged);
+
+        return result;
+    }
+
+    // Calls a managed classlib function to potentially get an unmanaged callback.
+    // Not for use with ObjectiveCMarshalTryGetTaggedMemory.
+    template <typename T>
+    T GetCallbackViaClasslibCallback(_In_ Object * object, _In_ ClasslibFunctionId id)
```

Comment:
I don't understand why this is templatized. This is a simple cast that could be performed at the callsite instead. I'm all for templates when they reduce code duplication, but this doesn't seem needed.

Reply:
Updated to not use templates.